rudderlabs/rudder-sdk-kotlin
# feat: process event in a sequential way

## Description


We've added the logic to process the events sequentially. Previously, we were processing events in a parallel way which can cause ordering guarantee issues.

## Type of change


## Implementation Details


### Analytics

We have added a channel with unlimited capacity in the Analytics class which will process the events in the background thread but in a sequential way.

## Checklist


## How to test?


- Make 1000 events and see the events in the Android console. 
- Send the above events to any server and note down the sequence.

## Breaking Changes


## Maintainers Checklist


## Screenshots (if applicable)


<img width="353" alt="image" src="https://github.com/user-attachments/assets/42e12ddc-e816-4f68-8fe1-2eb16f8c948c">


## Additional Context


- I tried a lot to add some unit tests to test the changes, but since Analytics is not designed that way, it's not possible to write an ordering guarantee unit test.
- I've also added a TODO to close the channel when SDK is either shutdown or the app is cleared from the memory.
